nodebuilder/header/metrics.go

```diff
@@ -0,0 +1,186 @@
+			"node.header.blackbox.request_duration",
+			instrument.WithDescription("duration of a single get header request"),
+		)
+	if err != nil {
+		return nil, err
+	}
+
+	requestSize, err := meter.
+		SyncInt64().
+		Histogram(
+			"node.header.blackbox.request_size",
+			instrument.WithDescription("size of a get header response"),
+		)
+	if err != nil {
+		return nil, err
+	}
+
+	blockTime, err := meter.
+		SyncInt64().
+		Histogram(
+			"node.header.blackbox.block_time",
+			instrument.WithDescription("test block time"),
+		)
+	if err != nil {
+		return nil, err
+	}
+
+	lastHeadTS := time.Now()
+	lastHead := int64(0)
+
+	bbinstrument := &blackBoxInstrument{
+		requestsNum,
+		requestDuration,
+		requestSize,
+		blockTime,
+		lastHeadTS,
+		lastHead,
+		next,
+	}
+
+	return bbinstrument, nil
+}
+
+// GetByHeight returns the ExtendedHeader at the given height, blocking
+// until header has been processed by the store or context deadline is exceeded.
+func (bbi *blackBoxInstrument) GetByHeight(ctx context.Context, height uint64) (*header.ExtendedHeader, error) {
+	now := time.Now()
+	requestID, err := misc.RandString(5)
+	if err != nil {
+		return nil, err
+	}
+
+	// defer recording the duration until the request has received a response and finished
+	defer func(ctx context.Context, begin time.Time) {
+		bbi.requestDuration.Record(
+			ctx,
+			time.Since(begin).Milliseconds(),
+		)
+	}(ctx, now)
+
+	// perform the actual request
+	eh, err := bbi.next.GetByHeight(ctx, height)
+	if err != nil {
+		// count the request and tag it as a failed one
+		bbi.requestsNum.Add(
+			ctx,
+			1,
+			attribute.String("request-id", requestID),
+			attribute.String("state", "failed"),
+		)
+		return eh, err
+	}
+
+	// other wise, count the request but tag it as a succeeded one
+	bbi.requestsNum.Add(
+		ctx,
+		1,
+		attribute.String("request-id", requestID),
+		attribute.String("state", "succeeded"),
+	)
+
+	// retrieve the binary format to get the size of the header
+	// TODO(@team): is ExtendedHeader.MarshalBinary() == ResponseSize?
+	// I am making this assumption for now
+	bin, err := eh.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+
+	// record the response size (extended header in this case)
+	bbi.requestSize.Record(
+		ctx,
+		int64(len(bin)),
+	)
+
+	return eh, err
+}
+
+// Head returns the ExtendedHeader of the chain head.
+func (bbi *blackBoxInstrument) Head(ctx context.Context) (*header.ExtendedHeader, error) {
+	return bbi.next.Head(ctx)
+}
+
+// Head returns the ExtendedHeader of the chain head.
+func (bbi *blackBoxInstrument) SyncerHead(ctx context.Context) (*header.ExtendedHeader, error) {
+	header, err := bbi.next.Head(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if header.RawHeader.Height > bbi.lastHead {
+		bbi.lastHead = header.RawHeader.Height
+		bbi.blockTime.Record(
+			ctx,
+			time.Since(bbi.lastHeadTS).Milliseconds(),
+			attribute.Int("height", int(header.RawHeader.Height)),
+			attribute.String("state", "failed"),
```

Comment:
why would this be state failed?

Reply:
weird, yeah probably a relic from debugging